Populate the bloom fulltext index on startup

chrisvest opened the pull request:
This builds on #9888, and together with #10014 should give us functional completeness, though we'd ideally like to only rebuild the index when it's strictly necessary, e.g. if relevant configurations have changed.

chrisvest committed:
Move fulltext index update application to a background thread but keep it synchronous (not eventually consistent... yet, anyway)

chrisvest committed:
The fulltext index is now populated from store scans on start-up

chrisvest committed:
Add a procedure for waiting for the fulltext index to finish populating e.g. after startup

chrisvest committed:
The fulltext index now deletes any existing index files from any prior starts of the database, in order to avoid configuration changes from creating persistent stale index entries

chrisvest committed:
License headers

chrisvest committed:
Fix checkstyle in new class

chrisvest committed:
Add a state check to avoid starting multiple fulltext applier threads, and increase the population transaction timeout to 10 hours to patiently wait for very long recovery times if necessary

chrisvest committed:
Add a test that verifies that the bloom fulltext index observes new data as it is committed

chrisvest committed:
Improve the multi-threaded write throughput of the bloom fulltext index add-on by about a factor four.

We now drain the work queue in the applier and apply the updates as one batch, and only refresh any given index once per batch.

chrisvest committed:
Use the availability guard to wait for the database to start before populating the fulltext index, since the populating transaction is otherwise likely to time out

chrisvest committed:
Fix checkstyle

chrisvest committed:
Put fulltext index add-on benchmarks in their own file and make them non-tests

ragadeeshu committed:
Add documents in batches for fulltext population

This makes population roughly twice as fast.

chrisvest committed:
Make the fulltext index add-on populator have an upper bound on its batch size, of 10.000 documents.

Previously it would create a list of documents for every node or relationship before writing them to Lucene. Since we source from all-nodes or all-relationships scans, these lists could become far too large to handle.

ragadeeshu committed:
Remove unused import in fulltext addon

ragadeeshu committed:
Fulltext addon population separate documents

This fixes a bug where population of more that one property would not
work, due to way to eager reuse of documents.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -65,13 +68,32 @@ public FulltextProvider( GraphDatabaseService db, Log log )
     public void close()
     {
         db.unregisterTransactionEventHandler( fulltextTransactionEventUpdater );
+        applier.stop();
```
Will it be impossible to stop while population is not completed right?

chrisvest replied:
As it stands, yeah. We have to pull the population out as an explicit life cycle phase, for it to work well with the recovery as well, if we allow clean shut-down mid-population.

MishaDemianenko replied:
ok

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -65,13 +68,32 @@ public FulltextProvider( GraphDatabaseService db, Log log )
+
+    public void awaitPopulation() throws IOException
+    {
+        applier.writeBarrier().await();
```
it would be enough and much simpler to keep a map of indexes with their statuses (2 indexes)

MishaDemianenko replied:
still open question right?

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextProvider.java:
```diff
@@ -65,13 +68,32 @@ public FulltextProvider( GraphDatabaseService db, Log log )
+        db.registerTransactionEventHandler( fulltextTransactionEventUpdater );
+    }
+
+    public void awaitPopulation() throws IOException
```
what is intended lifecycle of this call: who should call it? should w serve updates while indexes are not populated?

chrisvest replied:
I'll add some javadoc.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java:
```diff
@@ -87,77 +86,32 @@ public Object beforeCommit( TransactionData data ) throws Exception
-        //update node indices
         try
         {
+            List<BinaryLatch> completions = new ArrayList<>();
```
what do we gaining by generating lots independent small updates? why not batch them to at least one per index?

chrisvest replied:
The applier batches as it consumes from the queue, but you want to put the loop into the applier? We can do that.

MishaDemianenko replied:
yeah basically generate only one update that ges to applier queue at least per index per commit

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java:
```diff
@@ -0,0 +1,365 @@
+                    Map<String,Object> allProperties = state.get( entityId );
+                    if ( allProperties == null || allProperties.isEmpty() )
+                    {
+                        index.getIndexWriter().deleteDocuments( newTermForChangeOrRemove( entityId ) );
```
this thing still creates index writer each time

chrisvest replied:
The writer is cached in the `index` now.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java:
```diff
@@ -0,0 +1,365 @@
+        Pair<WritableFulltext,BinaryLatch> applyUpdateAndReturnIndex() throws IOException;
+    }
+
+    private static class ApplierThread extends Thread
```
We create our of thread here, maybe use Neo4jJobScheduler instead?

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java:
```diff
@@ -0,0 +1,365 @@
+        public void run()
+        {
+            waitForDatabaseToBeAvailable();
+            Set<WritableFulltext> refreshableSet = Collections.newSetFromMap( new IdentityHashMap<>() );
```
why set can't be simple HashSet, for example? why do we need identity hash map in here?

MishaDemianenko replied:
still open question here

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java:
```diff
@@ -0,0 +1,365 @@
+            {
+                isAvailable = availabilityGuard.isAvailable( 100 );
+            }
+            while ( !isAvailable && !availabilityGuard.isShutdown() );
```
not sure how useful `!availabilityGuard.isShutdown()` here since available it will be only in case if it's not shut down already and in the worst case, it can be shut down a moment after you check that shutdown flag anyway

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java:
```diff
@@ -0,0 +1,365 @@
+import static org.neo4j.kernel.api.impl.fulltext.LuceneFulltextDocumentStructure.documentRepresentingProperties;
+import static org.neo4j.kernel.api.impl.fulltext.LuceneFulltextDocumentStructure.newTermForChangeOrRemove;
+
+class FulltextUpdateApplier
```
why do we use only single threaded population here?  take a look how `BatchingMultipleIndexPopulator` does it for example if you did not yet

chrisvest replied:
It's missing infrastructure for indexing relationships, so it's not an integration we can do at this point.

MishaDemianenko replied:
hm you can still start multiple threads (maybe not in thins PR) to perfrom index population to make it times faster then from single thread

chrisvest replied:
Yeah, I've created a card for it.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/LuceneFulltextDocumentStructure.java:
```diff
@@ -62,6 +62,14 @@ static Field encodeValueField( String propertyKey, Value value )
         return LuceneFulltextFieldEncoding.encodeField( propertyKey, value );
     }
 
+    public static Document documentForPopulation( long entityId, Map<String,Object> properties )
```
don't think you need it as well since our standard index population know how to overcome that limitation already

chrisvest replied:
It makes no observable difference to performance to build a list of capturing lambdas instead, which we'd have to do because the call needs to know the size of the iterator up front. So I'm not sure it's worth the trouble in this case.

MishaDemianenko replied:
it's more a question to have common way of doing things and avoid bicycles here and there that do the same thing but on their own special way

chrisvest replied:
Alright.

MishaDemianenko commented on enterprise/fulltext-addon/src/test/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomFulltextIndexBenchmarks.java:
```diff
@@ -0,0 +1,177 @@
+
+@Ignore( "These are rudimentary benchmarks, but implemented via the jUnit framework to make them easy to run " +
+         "from an IDE." )
+public class BloomFulltextIndexBenchmarks
```
Since it's a benchmark it should not live here at all, no reason to introduce dead code in the repo

chrisvest replied:
We'll delete them when we the architecture is more settled.

MishaDemianenko commented on enterprise/fulltext-addon/src/test/java/org/neo4j/kernel/api/impl/fulltext/LuceneFulltextUpdaterTest.java:
```diff
@@ -666,17 +686,85 @@ public void shouldNotReturnNonMatches() throws Exception
+            try ( ReadOnlyFulltext reader = provider.getReader( "nodes", NODES ) )
+            {
+
+                PrimitiveLongIterator hello = reader.query( "hello" );
```
it's super hard to understand what this test is actually testing because of zebras of asserts, can those be grouped extracted in some way to be able to understand whats going on without  👓  and 🍷

chrisvest replied:
Let's do that in #10079 where we've already made significant changes to this file.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java:
```diff
@@ -0,0 +1,365 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.neo4j.concurrent.BinaryLatch;
```
While it's understandable why you use latches I do not feel comfortable with that level of abstraction that escapes from this class, I think this should be simplified since we do not need that level of precision (per update) anyway and we can be quite relaxed in general about queue status

chrisvest replied:
I think it's an important architectural flexibility that we can submit updates asynchronously, and then wait for their completion. It has not yet been decided if we want the index to be eventually consistent or synchronous. We don't have to expose the concrete BinaryLatch, though. I can hide that behind an interface.

MishaDemianenko reviewed (changes requested):
The first round of review completed!
Would be great if we will update PR description a bit to describe what actually going on inside of this change request.
It's not really clear so far for me how clients should use provided procedures in terms of indexes availability and how we can guard them against miss use, and other questions here and there.

ragadeeshu committed:
Reuse index writers for fulltext addon

chrisvest committed:
Hide the BinaryLatch returned from the FulltextUpdateApplier behind an interface

chrisvest committed:
Start the fulltext index applier background thread via the JobScheduler

chrisvest committed:
Document the purpose of the FulltextProvider.awaitPopulation method.

chrisvest committed:
Remove unnecessary check for availability guard shutdown in the fulltext update applier

chrisvest commented:
@MishaDemianenko I think I've addressed all your comments.

MishaDemianenko commented:
@chrisvest @ragadeeshu couple of open questions still, most of the comments solved already by new commits

chrisvest committed:
Avoid allocating whole documents when populating the fulltext index add-on.

Allocate document factory closures instead, so a single document can be reused.

chrisvest commented:
@MishaDemianenko Removed the document allocation.

chrisvest committed:
This does not need to be an identity hash set since the WritableFulltext objects use default identity based equals and hashcode anyway.
